### PR TITLE
Snapshots documentation

### PR DESCRIPTION
**Pull request by geordieintheshellcode**
Documentation for the Ondat snapshots feature. This is split into two parts:
1. A concepts page, giving an overview of the feature
2. A operations page, giving a detailed walk-through of how to use the feature.

### CODE REVIEW

File: docs/concepts/snapshots.md [hunk 1]
```diff
@@ -5,12 +5,67 @@ linkTitle: Ondat Snapshots
...
+The Kubernetes [Volume
+Snapshots](https://kubernetes.io/docs/concepts/storage/volume-snapshots/)
+feature provides users with a set of custom resource definitions (CRD) and APIs
+to create and manage volume snapshots. Storage providers can then implement the
+necessary CSI APIs to integrate with this feature. This is exactly what we’ve
+done at Ondat. Additional tools, like Kasten K10, can then be utilised to
+orchestrate and automate snapshotting, backups and restores. 
+
+Please see the [Backups and restores with Kasten
+K10](/docs/operations/backups-and-restores-with-kastenk10.md) for a full
```
Comment (cannischan) [inline docs/concepts/snapshots.md:53]: remove .md to make the url work

File: docs/concepts/snapshots.md [hunk 2]
```diff
@@ -5,12 +5,67 @@ linkTitle: Ondat Snapshots
...
+walk through.
+
+Note: the Ondat Snapshot feature is not fully CSI compliant. As such the feature
```
Comment (cannischan) [inline docs/concepts/snapshots.md:56]: Replace with:

> ⚠️ The Ondat Snapshot feature is not fully CSI compliant. As such the feature

File: docs/concepts/snapshots.md [hunk 3]
```diff
@@ -5,12 +5,67 @@ linkTitle: Ondat Snapshots
...
-> ⚠️ This feature is currently available on `2.8.0-beta` - if you would like to gain early access to the product, speak with our experts on hello@ondat.io.
+1. The feature has been designed to work with Kasten K10 only. This is not a
+fully CSI compliant implementation of the spec.
+1. Restoring via Kasten 10 from a “local snapshot” is not supported with the
+Ondat Snapshot feature. Users may only restore applications using a Kasten K10
+“External backup”. 
+1. Snapshotting RWX volumes is not supported. This is because it is next to
+impossible to ensure that a NFS mounted volume is in a suitable state for
+snapshotting. For RWX volumes, the user only has access to the filesystem on
+the NFS client. It is not possible to run `fsfreeze` on this mountpoint - NFS
```
Comment (cannischan) [inline docs/concepts/snapshots.md:71]: "It is not possible to run fsfreeze..." -> this should be a separate bullet?

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+title: "Backup and restores with Ondat Snapshots and Kasten K10"
+linkTitle: Backup and restores with Ondat Snapshots and Kasten K10
+---
+
+# Overview
+
+This guide will walk you through how to use the Ondat Snapshots feature to
+backup and restore your Kubernetes applications using Kasten K10. Before
+starting please consult the [Snapshots
+Concepts](/docs/concepts/snapshots.md) page for an overview of the feature.
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md:11]: remove .md for the url to work

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+    1. Creating a backup policy
+    1. Adding pre/post-snapshot hooks to quiesce the application/filesystem
+    1. Manually running a backup job
+    1. Restoring an application from a backup
+
+# 1: Installing Ondat
+
+Ondat Snapshots were introduced in v2.8.0. If you are installing Ondat for the
+first time then please follow the instructions
+[here](https://docs.ondat.io/docs/install/). If you are upgrading an existing
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md:30]: update url to `/docs/install/`

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+Ondat deployment then please follow the instructions
+[here](https://docs.ondat.io/docs/upgrade/upgrade/).
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md:32]: update url to `/docs/upgrade/upgrade/`

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+
+# 4: Installing Kasten K10
+
+The next step is to install Kasten K10 on your cluster. Instructions for how to do this can
+be found [here](https://docs.kasten.io/latest/install/install.html).
+
+The remainder of this walk through will assume you have access to the Kasten K10 UI. You
+can install it following the instructions
+[here](https://docs.kasten.io/latest/access/dashboard.html). Everything we do in the
+following steps may be done via `kubectl` and the command line, however, but is not shown
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md]: ...and the command line, **however this is not** shown in this guide

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+specific management tasks via yaml based configuration. Kasten K10 can leverage
+these blueprints to perform various actions pre and post-snapshot.
+
+We’ll use the following blueprint to quiesce the filesystem before taking a
+snapshot and unquiesce it after taking the snapshot.
+
+Note: we don’t perform any application level quiescing here as it’s not required
+for our toy application.
+
+Note: for statefulsets the blueprint below can be used as a baseline, but change
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md:234]: Replace Note: with >

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+        - -o
+        - pipefail
+        - -o
+        - xtrace
+        - -c
+        - |
+          fsfreeze --unfreeze /mnt
+```
+
+Note: if you have installed kasten into a namespace other than `kasten-io` then
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md:287]: Replace Note: to >

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+```
+
+## 5.3: Setting up a backup policy
+
+Ensure you have the Kasten K10 dashboard installed
+(see [here](https://docs.kasten.io/latest/access/dashboard.html)). It’s possible
+to do the steps in this section via `kubectl` and the command line. Please see the
+Kasten K10 documentation for how to do this.
+
+Browse to the "Policies" page and click "Create New Policy":
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md:308]: Go to the "Policies" page

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+![EditPolicy1](/images/docs/operations/backups/image2.png)
+
+Under "Select Applications", check the "By Name" checkbox and search for the
+namespace "ondat-test".
+
+![EditPolicy2](/images/docs/operations/backups/image3.png)
+
+Leave everything else as is then click `Create Policy` to create the policy.
+
+Note: do not try to set the `Pre and Post-Snapshot Action Hooks` in the `Advanced Settings`
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md:328]: Replace Note: with >⚠️

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+
+## 5.5: Restoring an application from a backup
+
+Let’s emulate a disaster recovery scenario, by deleting our deployment:
+
+```
+$ kubectl delete -n ondat-test deployments.apps myapp-deployment
+```
+
+Now let’s restore it. First browse to the dashboard and select "Applications".
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md:369]: First go to the dashboard

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+
+## Known issues and gotchas
+
+### Deletion of VolumeSnapshot
+
+Ideally the Kubernetes `VolumeSnapshot` object should be deleted as soon as the
+application has been backed up externally. There’s no reason to keep the snapshot
+as it can’t be meaningfully used. Furthermore, it takes up space on the local node
+and can cause a slight performance degradation for IO to the parent volume. At the
+moment, there is no good way to automate this procedure. Consequently user's should:
```
Comment (cannischan) [inline docs/operations/backups-and-restores-with-kastenk10.md:429]: Consequently **users** should

**Review by cannischan — CHANGES_REQUESTED** [review]
Some comments on formatting/typos

File: docs/concepts/snapshots.md (hunk 1 above)
Comment (geordieintheshellcode) [reply docs/concepts/snapshots.md:53]: Done

File: docs/concepts/snapshots.md (hunk 2 above)
Comment (geordieintheshellcode) [reply docs/concepts/snapshots.md:56]: Done

File: docs/concepts/snapshots.md (hunk 3 above)
Comment (geordieintheshellcode) [reply docs/concepts/snapshots.md:71]: Nope it's the same sentence.

File: docs/operations/backups-and-restores-with-kastenk10.md (hunk 4 above)
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md:11]: Done

File: docs/operations/backups-and-restores-with-kastenk10.md (hunk 5 above)
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md:30]: Done

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+install these by default. To install the `VolumeSnapshots`,
+`VolumeSnapshotContents` and `VolumeSnapshotClasses` CRDs run the following:
+
+```
+$ kubectl create -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/release-6.0/client/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml
+$ kubectl create -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/release-6.0/client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml
+$ kubectl create -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/release-6.0/client/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml
+```
+
+Now let's need to install the `snapshot-controller`. The
```
[the remark this answers is not in the document]
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md:47]: Done

File: docs/operations/backups-and-restores-with-kastenk10.md (hunk 7 above)
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md]: Done

File: docs/operations/backups-and-restores-with-kastenk10.md [hunk 15]
```diff
@@ -0,0 +1,445 @@
...
+[Kanister blueprint](https://kanister.io/) in conjunction with K10’s
+[execution hooks](https://docs.kasten.io/latest/kanister/hooks.html) feature.
+Kanister is an open source framework which allows users to automate application
+specific management tasks via yaml based configuration. Kasten K10 can leverage
+these blueprints to perform various actions pre and post-snapshot.
+
+We’ll use the following blueprint to quiesce the filesystem before taking a
+snapshot and unquiesce it after taking the snapshot.
+
+Note: we don’t perform any application level quiescing here as it’s not required
```
[the remark this answers is not in the document]
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md:231]: Done

File: docs/operations/backups-and-restores-with-kastenk10.md (hunk 8 above)
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md:234]: Done

File: docs/operations/backups-and-restores-with-kastenk10.md (hunk 9 above)
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md:287]: Done

File: docs/operations/backups-and-restores-with-kastenk10.md (hunk 10 above)
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md:308]: Done

File: docs/operations/backups-and-restores-with-kastenk10.md (hunk 12 above)
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md:369]: Done

File: docs/operations/backups-and-restores-with-kastenk10.md (hunk 13 above)
Comment (geordieintheshellcode) [reply docs/operations/backups-and-restores-with-kastenk10.md:429]: Done

**Review by cannischan — DISMISSED** [review]

**Review by cannischan — APPROVED** [review]

**Pull request merged**